src/index.ts

```diff
@@ -0,0 +1,188 @@
+import { parse } from 'graphql';
+import { Location, DocumentNode, DefinitionNode } from 'graphql/language';
+
+// Strip insignificant whitespace
+// Note that this could do a lot more, such as reorder fields etc.
+function normalize(text: string): string {
+  return text.replace(/[\s,]+/g, ' ').trim();
+}
+
+type DocCache = { [docString: string]: DocumentNode };
+let docCache: DocCache = {};
+
+interface FragmentSourceMap {
+  [framentName: string]: {
+    [source: string]: boolean;
+  };
+}
+
+// A map fragmentName -> [normalized source]
+let fragmentSourceMap: FragmentSourceMap = {};
+
+function cacheKeyFromLoc(loc?: Location): string {
```

Comment:
just for the sake of the translation, i dont think we should be changing the behavior of the individual functions just yet. any potential improvements _should_ be marked, but im afraid that this is going to have a cascading effect on existing behavior that we can't keep track of.

Reply:
here's the source:
> function cacheKeyFromLoc(loc) {
  return normalize(loc.source.body.substring(loc.start, loc.end));
}

Reply:
It is the same as the source afaik?